src/Application/Actions/Donations/Update.php

```diff
@@ -162,15 +162,18 @@ protected function action(Request $request, Response $response, array $args): Re
                     try {
                         $this->donationService->cancel($donation);
                     } catch (DonationAlreadyFinalised $e) {
+                        $this->entityManager->rollback();
```

Comment:
had to move around quite a bit of transaction stuff. Would like to try and keep to a rule of always having any calls related to the same transaction in the same function - I made a mistake before by moving code with a transaction commit out of here into a service class while keeping the transaction start and rollbacks here.

Could force that by using \MatchBot\Application\Persistence\RetrySafeEntityManager::wrapInTransaction instead but that's doing more stuff and is a bigger change, especially since our implementation of that has retrying built-in.

Reply:
Yeah, I think there's a trade off between those for sure. I like `wrapInTransaction()` better conceptually, but above a certain size of transaction I feel like it makes the code harder to follow again.

Am I right in thinking these changes reverse what I found about `DonationService::cancel()` needing an open transaction? If so then my comment on the fn from yesterday should go too to avoid adding more confusion.

Reply:
Yes, thanks for spotting. Done in 9c4db075

Reply:
Not about to try to build it but I guess it might be nice if there was an interface where you could do something like:

```
$transaction = $entityManager->beginTransaction(); // (or $this->transaction = ...)
/// lots of code
$wasSucessful ? $transaction->commit(); $transaction->rollback();
 ```